be1-go/validation/schema_validator.go

```diff
@@ -0,0 +1,93 @@
+package validation
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"student20_pop/message"
+
+	"github.com/xeipuuv/gojsonschema"
+	"golang.org/x/xerrors"
+)
+
+type SchemaValidator struct {
+	schemas map[string]*gojsonschema.Schema
+}
+
+type schema struct {
+	name string
+	path string
+}
+
+const (
+	GenericMsgSchema string = "genericMsgSchema"
+	DataSchema       string = "dataSchema"
+)
+
+// RegisterSchema adds a JSON schema specified by the path
+func (s *SchemaValidator) RegisterSchema(schema schema) error {
+	msgLoader := gojsonschema.NewReferenceLoader(schema.path)
+	msgSchema, err := gojsonschema.NewSchema(msgLoader)
+	if err != nil {
+		return xerrors.Errorf("failed to load the json schema with path `%s`: %v", schema.path, err)
+	}
+
+	s.schemas[schema.name] = msgSchema
+	return nil
+}
+
+func (s *SchemaValidator) VerifyJson(byteMessage []byte, schemaName string) error {
+	// Validate the Json "byteMessage" with a schema
+	messageLoader := gojsonschema.NewBytesLoader(byteMessage)
+	resultErrors, err := s.schemas[schemaName].Validate(messageLoader)
+	if err != nil {
+		return &message.Error{
+			Code:        -1,
+			Description: err.Error(),
+		}
+	}
+
+	errorsList := resultErrors.Errors()
+	descriptionErrors := ""
+	// Concatenate all error descriptions
+	for index, e := range errorsList {
+		descriptionErrors += fmt.Sprintf(" (%d) %s", index+1, e.Description())
+	}
+
+	if len(errorsList) > 0 {
+		return &message.Error{
+			Code:        -1,
+			Description: descriptionErrors,
+		}
+	}
+
+	return nil
+}
+
+func NewSchemaValidator() (*SchemaValidator, error) {
+	// Import the Json schemas defined in the protocol section
+	protocolPath, err := filepath.Abs("../protocol")
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load the path for the json schemas: %v", err)
+	}
+	protocolPath = "file://" + protocolPath
+
+	genericMsgSchema := schema{GenericMsgSchema, protocolPath + "/genericMessage.json"}
```

Comment:
Is the path compatible across all environments? Specially Windows?

Reply:
Also, please use the field names while declaring a struct

Reply:
The windows path had `\\` instead of `/`, I just replaced all `\\` by `/`.